server/src/transport/ws.rs

```diff
@@ -276,13 +278,25 @@ pub(crate) async fn background_task<L: Logger>(sender: Sender, mut receiver: Rec
 
 	tokio::pin!(stopped);
 
-	let result = loop {
-		data.clear();
+	let ws_stream = futures_util::stream::unfold((Vec::new(), receiver), |(mut data, mut receiver)| async {
+		match receiver.receive(&mut data).await {
+			Ok(soketto::Incoming::Data(_)) => Some((Ok(Incoming::Data(std::mem::take(&mut data))), (data, receiver))),
+			Ok(soketto::Incoming::Pong(_)) => Some((Ok(Incoming::Pong), (data, receiver))),
+			Ok(soketto::Incoming::Closed(_)) | Err(SokettoError::Closed) => None,
+			// The closing reason is already logged by `soketto` trace log level.
+			// Return the `Closed` error to avoid logging unnecessary warnings on clean shutdown.
+			Err(e) => Some((Err(e), (data, receiver))),
+		}
+	});
```

Comment:
So if I understand this:
- When we see `Incoming::Data`, we take the data from the buffer and hand back `Incoming::Data(owned_vec)`.
- When we see `Incoming::Pong`, we leave the buffer alone (might it have data in it now or not?). This same buffer will be written to again next time (so if there is data in it, I guess it'll be wrong?).
- Exit on any errors.

I guess I wonder whether it would be better to have:

```rust
let ws_stream = futures_util::stream::unfold(receiver, |mut receiver| async {
	let mut data = Vec::new();
	match receiver.receive(&mut data).await {
		Ok(soketto::Incoming::Data(_)) => Some((Ok(Incoming::Data(data)), receiver)),
		Ok(soketto::Incoming::Pong(_)) => Some((Ok(Incoming::Pong), receiver)),
		Ok(soketto::Incoming::Closed(_)) | Err(SokettoError::Closed) => None,
		// The closing reason is already logged by `soketto` trace log level.
		// Return the `Closed` error to avoid logging unnecessary warnings on clean shutdown.
		Err(e) => Some((Err(e), receiver)),
	}
});
```

Unless the whole point is that we need to keep the buffer around because data might be written during eg a Pong that is important when we get to `Incoming::Data` or something? (I'd need to look at soketto I guess!)

Reply:
yeah your suggestion is simpler so why not :)